File: TimskiMilenici/src/main/java/com/example/timskimilenici/services/OwnerAnalyticsService.java
package com.example.timskimilenici.services;

import com.example.timskimilenici.repositories.BookingRepository;
import com.example.timskimilenici.repositories.OrderRepository;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.sql.Date;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.temporal.ChronoUnit;
import java.util.List;

@Service
public class OwnerAnalyticsService {

    private final BookingRepository bookingRepository;
    private final OrderRepository orderRepository;

    public OwnerAnalyticsService(BookingRepository bookingRepository, OrderRepository orderRepository) {
        this.bookingRepository = bookingRepository;
        this.orderRepository = orderRepository;
    }

    public record OverviewPoint(
            LocalDate day,
            long bookings,
            long completed,
            long cancelled,
            BigDecimal revenue
    ) {}

    public record OverviewSummary(
            Double completionRate,
            BigDecimal averageBookingValue,
            BigDecimal priorPeriodRevenue,
            Double revenueChangePercent
    ) {}

    public record OverviewResult(
            List<OverviewPoint> points,
            long totalBookings,
            long totalCompleted,
            long totalCancelled,
            BigDecimal totalRevenue,
            OverviewSummary summary
    ) {}

    public record ProductSalesRow(
            Long productId,
            String productName,
            long unitsSold,
            long ordersCount,
            BigDecimal revenue
    ) {}

    public record PeakTimeBucket(
            int dayOfWeek,
            int hour,
            long bookings
    ) {}

    public record SpecialOfferRow(
            Long itemId,
            String itemName,
            /** "service" or "product". */
            String itemType,
            Long businessId,
            String businessName,
            BigDecimal basePrice,
            BigDecimal promotionPrice,
            /** Current discount percent (0-100). */
            Double discountPercent,
            /** Bookings/orders in this period. */
            long usageCount,
            /** Revenue earned during this period from this promoted item. */
            BigDecimal promotedRevenue
    ) {}

    public record SpecialOffersResult(
            int activeOfferCount,
            long totalUsageCount,
            BigDecimal totalPromotedRevenue,
            /** Average discount percent across active offers (0-100), or null when no offers. */
            Double averageDiscountPercent,
            /** Top offers sorted by usage desc, capped by caller (default 5). */
            List<SpecialOfferRow> topOffers,
            /** Every active offer for the owner in this period. */
            List<SpecialOfferRow> offers
    ) {}

    public OverviewResult getOverview(Long ownerId, LocalDate fromDate, LocalDate toDate, Long businessId) {
        OverviewResult current = buildOverviewCore(ownerId, fromDate, toDate, businessId);
        List<OverviewPoint> points = current.points();
        long totalBookings = current.totalBookings();
        long totalCompleted = current.totalCompleted();
        long totalCancelled = current.totalCancelled();
        BigDecimal totalRevenue = current.totalRevenue();

        Double completionRate = totalBookings > 0
                ? (double) totalCompleted / (double) totalBookings
                : null;
        BigDecimal averageBookingValue = totalBookings > 0
                ? totalRevenue.divide(BigDecimal.valueOf(totalBookings), 2, RoundingMode.HALF_UP)
                : null;

        long inclusiveDays = ChronoUnit.DAYS.between(fromDate, toDate) + 1;
        LocalDate priorPeriodEnd = fromDate.minusDays(1);
        LocalDate priorPeriodStart = priorPeriodEnd.minusDays(inclusiveDays - 1);
        OverviewResult prior = buildOverviewCore(ownerId, priorPeriodStart, priorPeriodEnd, businessId);
        BigDecimal priorRevenue = prior.totalRevenue();
        Double revenueChangePercent = null;
        if (priorRevenue.compareTo(BigDecimal.ZERO) > 0) {
            revenueChangePercent = totalRevenue.subtract(priorRevenue)
                    .divide(priorRevenue, 4, RoundingMode.HALF_UP)
                    .multiply(BigDecimal.valueOf(100))
                    .doubleValue();
        } else if (totalRevenue.compareTo(BigDecimal.ZERO) > 0) {
            revenueChangePercent = null;
        } else {
            revenueChangePercent = 0.0;
        }

        OverviewSummary summary = new OverviewSummary(
                completionRate,
                averageBookingValue,
                priorRevenue,
                revenueChangePercent
        );

        return new OverviewResult(points, totalBookings, totalCompleted, totalCancelled, totalRevenue, summary);
    }

    private OverviewResult buildOverviewCore(Long ownerId, LocalDate fromDate, LocalDate toDate, Long businessId) {
        LocalDateTime from = fromDate.atStartOfDay();
        LocalDateTime to = toDate.atTime(LocalTime.MAX);

        List<Object[]> rows = bookingRepository.aggregateOverviewByOwnerAndDate(ownerId, from, to, businessId);
        List<OverviewPoint> points = rows.stream()
                .map(r -> {
                    LocalDate day = ((Date) r[0]).toLocalDate();
                    BigDecimal revenue = r[4] instanceof BigDecimal ? (BigDecimal) r[4] : BigDecimal.ZERO;
                    return new OverviewPoint(
                            day,
                            ((Number) r[1]).longValue(),
                            ((Number) r[2]).longValue(),
                            ((Number) r[3]).longValue(),
                            revenue
                    );
                })
                .toList();

        long totalBookings = points.stream().mapToLong(OverviewPoint::bookings).sum();
        long totalCompleted = points.stream().mapToLong(OverviewPoint::completed).sum();
        long totalCancelled = points.stream().mapToLong(OverviewPoint::cancelled).sum();
        BigDecimal totalRevenue = points.stream()
                .map(OverviewPoint::revenue)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        OverviewSummary summary = new OverviewSummary(null, null, null, null);
        return new OverviewResult(points, totalBookings, totalCompleted, totalCancelled, totalRevenue, summary);
    }

    public List<ProductSalesRow> getProductSales(Long ownerId, LocalDate fromDate, LocalDate toDate, Long businessId) {
        LocalDateTime from = fromDate.atStartOfDay();
        LocalDateTime to = toDate.atTime(LocalTime.MAX);

        List<Object[]> rows = orderRepository.aggregateProductSalesByOwner(ownerId, from, to, businessId);
        return rows.stream()
                .map(r -> new ProductSalesRow(
                        ((Number) r[0]).longValue(),
                        (String) r[1],
                        ((Number) r[2]).longValue(),
                        ((Number) r[3]).longValue(),
                        (BigDecimal) r[4]
                ))
                .toList();
    }

    public SpecialOffersResult getSpecialOffers(Long ownerId, LocalDate fromDate, LocalDate toDate, Long businessId) {
        LocalDateTime from = fromDate.atStartOfDay();
        LocalDateTime to = toDate.atTime(LocalTime.MAX);

        List<SpecialOfferRow> offers = new java.util.ArrayList<>();

        List<Object[]> serviceRows = bookingRepository.aggregatePromotedServices(ownerId, from, to, businessId);
        for (Object[] r : serviceRows) {
            BigDecimal base = r[4] instanceof BigDecimal ? (BigDecimal) r[4] : BigDecimal.ZERO;
            BigDecimal promo = r[5] instanceof BigDecimal ? (BigDecimal) r[5] : BigDecimal.ZERO;
            BigDecimal revenue = r[8] instanceof BigDecimal ? (BigDecimal) r[8] : BigDecimal.ZERO;
            offers.add(new SpecialOfferRow(
                    ((Number) r[0]).longValue(),
                    (String) r[1],
                    "service",
                    ((Number) r[2]).longValue(),
                    (String) r[3],
                    base,
                    promo,
                    computeDiscountPercent(base, promo),
                    ((Number) r[6]).longValue(),
                    revenue
            ));
        }

        List<Object[]> productRows = orderRepository.aggregatePromotedProducts(ownerId, from, to, businessId);
        for (Object[] r : productRows) {
            BigDecimal base = r[4] instanceof BigDecimal ? (BigDecimal) r[4] : BigDecimal.ZERO;
            BigDecimal promo = r[5] instanceof BigDecimal ? (BigDecimal) r[5] : BigDecimal.ZERO;
            BigDecimal revenue = r[8] instanceof BigDecimal ? (BigDecimal) r[8] : BigDecimal.ZERO;
            offers.add(new SpecialOfferRow(
                    ((Number) r[0]).longValue(),
                    (String) r[1],
                    "product",
                    ((Number) r[2]).longValue(),
                    (String) r[3],
                    base,
                    promo,
                    computeDiscountPercent(base, promo),
                    ((Number) r[7]).longValue(),
                    revenue
            ));
        }

        offers.sort((a, b) -> Long.compare(b.usageCount(), a.usageCount()));

        long totalUsage = offers.stream().mapToLong(SpecialOfferRow::usageCount).sum();
        BigDecimal totalRevenue = offers.stream()
                .map(SpecialOfferRow::promotedRevenue)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        Double avgDiscount = null;
        if (!offers.isEmpty()) {
            double sum = offers.stream()
                    .map(SpecialOfferRow::discountPercent)
                    .filter(java.util.Objects::nonNull)
                    .mapToDouble(Double::doubleValue)
                    .sum();
            avgDiscount = sum / offers.size();
        }

        List<SpecialOfferRow> top = offers.stream().limit(5).toList();

        return new SpecialOffersResult(
                offers.size(),
                totalUsage,
                totalRevenue,
                avgDiscount,
                top,
                offers
        );
    }

    private static Double computeDiscountPercent(BigDecimal base, BigDecimal promo) {
        if (base == null || promo == null) return null;
        if (base.compareTo(BigDecimal.ZERO) <= 0) return null;
        BigDecimal diff = base.subtract(promo);
        return diff.multiply(BigDecimal.valueOf(100))
                .divide(base, 1, RoundingMode.HALF_UP)
                .doubleValue();
    }

    public List<PeakTimeBucket> getPeakTimes(Long ownerId, LocalDate fromDate, LocalDate toDate, Long businessId) {
        LocalDateTime from = fromDate.atStartOfDay();
        LocalDateTime to = toDate.atTime(LocalTime.MAX);

        List<Object[]> rows = bookingRepository.aggregatePeakTimes(ownerId, from, to, businessId);
        return rows.stream()
                .map(r -> new PeakTimeBucket(
                        ((Number) r[0]).intValue(),
                        ((Number) r[1]).intValue(),
                        ((Number) r[2]).longValue()
                ))
                .toList();
    }
}

File: frontend/src/features/analytics/services/ownerAnalyticsService.ts
import { getStoredToken } from '../../auth/utils/tokenStorage';

const API_URL = 'http://localhost:8080/api/owner/analytics';

const getAuthToken = () => getStoredToken();

export interface OverviewPoint {
  day: string;
  bookings: number;
  completed: number;
  cancelled: number;
  revenue: number;
}

export interface OverviewSummary {
  completionRate: number | null;
  averageBookingValue: number | null;
  priorPeriodRevenue: number;
  revenueChangePercent: number | null;
}

export interface OverviewResult {
  points: OverviewPoint[];
  totalBookings: number;
  totalCompleted: number;
  totalCancelled: number;
  totalRevenue: number;
  /** Present when API returns extended overview; omitted on older backends */
  summary?: OverviewSummary;
}

export interface ProductSalesRow {
  productId: number;
  productName: string;
  unitsSold: number;
  ordersCount: number;
  revenue: number;
}

export interface PeakTimeBucket {
  dayOfWeek: number;
  hour: number;
  bookings: number;
}

export interface SpecialOfferRow {
  itemId: number;
  itemName: string;
  itemType: 'service' | 'product';
  businessId: number;
  businessName: string;
  basePrice: number;
  promotionPrice: number;
  discountPercent: number | null;
  usageCount: number;
  promotedRevenue: number;
}

export interface SpecialOffersResult {
  activeOfferCount: number;
  totalUsageCount: number;
  totalPromotedRevenue: number;
  averageDiscountPercent: number | null;
  topOffers: SpecialOfferRow[];
  offers: SpecialOfferRow[];
}

const authHeaders = (): Record<string, string> => {
  const token = getAuthToken();
  const h: Record<string, string> = {};
  if (token) h.Authorization = `Bearer ${token}`;
  return h;
};

interface AnalyticsQueryParams {
  businessId?: number | null;
  from?: string;
  to?: string;
}

const buildQuery = (params?: AnalyticsQueryParams) => {
  if (!params) return '';
  const search = new URLSearchParams();
  if (params.businessId) search.set('businessId', String(params.businessId));
  if (params.from) search.set('from', params.from);
  if (params.to) search.set('to', params.to);
  const qs = search.toString();
  return qs ? `?${qs}` : '';
};

export const ownerAnalyticsService = {
  getOverview: async (params?: AnalyticsQueryParams): Promise<OverviewResult> => {
    const res = await fetch(`${API_URL}/overview${buildQuery(params)}`, {
      headers: { 'Content-Type': 'application/json', ...authHeaders() },
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.message || 'Failed to load overview analytics');
    }
    return await res.json();
  },

  getProductSales: async (params?: AnalyticsQueryParams): Promise<ProductSalesRow[]> => {
    const res = await fetch(`${API_URL}/services${buildQuery(params)}`, {
      headers: { 'Content-Type': 'application/json', ...authHeaders() },
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.message || 'Failed to load service analytics');
    }
    return await res.json();
  },

  getPeakTimes: async (params?: AnalyticsQueryParams): Promise<PeakTimeBucket[]> => {
    const res = await fetch(`${API_URL}/peak-times${buildQuery(params)}`, {
      headers: { 'Content-Type': 'application/json', ...authHeaders() },
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.message || 'Failed to load peak time analytics');
    }
    return await res.json();
  },

  getSpecialOffers: async (params?: AnalyticsQueryParams): Promise<SpecialOffersResult> => {
    const res = await fetch(`${API_URL}/special-offers${buildQuery(params)}`, {
      headers: { 'Content-Type': 'application/json', ...authHeaders() },
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.message || 'Failed to load special offers analytics');
    }
    return await res.json();
  },
};

